File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/PriceSizeLadder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Cache
{
    /// <summary>
    /// A price size ladder with copy on write snapshot
    /// </summary>
    public class PriceSizeLadder
    {
        public static readonly IComparer<double> BACK_ORDER = new ReverseComparer<double>(Comparer<double>.Default);
        public static readonly IComparer<double> LAY_ORDER = Comparer<double>.Default;

        /// <summary>
        /// Dictionary of price to PriceSize.
        /// </summary>
        private readonly SortedDictionary<double, PriceSize> _priceToSize;
        private IList<PriceSize> _snap = PriceSize.EmptyList;


        public static PriceSizeLadder NewBack()
        {
            return new PriceSizeLadder(BACK_ORDER);
        }

        public static PriceSizeLadder NewLay()
        {
            return new PriceSizeLadder(LAY_ORDER);
        }

        private PriceSizeLadder(IComparer<double> comparer)
        {
            _priceToSize = new SortedDictionary<double, PriceSize>(comparer);
        }

        public IList<PriceSize> OnPriceChange(bool isImage, List<List<double?>> prices)
        {
            if (isImage)
            {
                //initial image - so clear cache
                _priceToSize.Clear();
            }
            if (prices != null)
            {
                //changes to apply
                foreach (List<double?> price in prices)
                {
                    PriceSize priceSize = new PriceSize(price);
                    if (priceSize.Size == 0.0)
                    {
                        //zero signifies remove
                        _priceToSize.Remove(priceSize.Price);
                    }
                    else
                    {
                        //update price
                        _priceToSize[priceSize.Price] = priceSize;
                    }
                }
            }
            if (isImage || prices != null)
            {
                //update snap on image or if we had cell changes
                _snap = new List<PriceSize>(_priceToSize.Values);
            }
            return _snap;
        }

    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/util/PriceSizeLadder.java
package com.betfair.esa.client.cache.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class PriceSizeLadder {
    private final Map<Double, PriceSize> priceToSize;
    private List<PriceSize> snap = Collections.emptyList();

    public static PriceSizeLadder newBack() {
        return new PriceSizeLadder(Comparator.reverseOrder());
    }

    public static PriceSizeLadder newLay() {
        return new PriceSizeLadder(Comparator.naturalOrder());
    }

    private PriceSizeLadder(Comparator<Double> comparator) {
        priceToSize = new TreeMap<>(comparator);
    }

    public List<PriceSize> onPriceChange(boolean isImage, List<List<Double>> prices) {
        if (isImage) {
            priceToSize.clear();
        }
        if (prices != null) {
            for (List<Double> price : prices) {
                PriceSize priceSize = new PriceSize(price);
                if (priceSize.getSize() == 0.0) {
                    priceToSize.remove(priceSize.getPrice());
                } else {
                    priceToSize.put(priceSize.getPrice(), priceSize);
                }
            }
        }
        if (isImage || prices != null) {
            // update snap on image or if we had cell changes
            snap = new ArrayList<>(priceToSize.values());
        }
        return snap;
    }

    @Override
    public String toString() {
        return "{" + priceToSize.values() + '}';
    }
}
